File: julia/all-your-base/all-your-base.jl
function all_your_base(digits, base_in, base_out)
    if base_in < 2 || base_out < 2
        throw(DomainError("base must be ≥ 2"))
    end

    # convert input digits to a decimal number
    value = 0
    for d in digits
        (0 ≤ d < base_in) || throw(DomainError("digit out of range"))
        value = value * base_in + d
    end

    # express decimal value as digits of output base
    result = []
    while true
        value, digit = divrem(value, base_out)
        pushfirst!(result, digit)
        if value == 0
            break
        end
    end
    result
end

File: elixir/all-your-base/lib/all_your_base.ex
defmodule AllYourBase do
  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """

  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(_, _, obase) when obase < 2, do: {:error, "output base must be >= 2"}
  def convert(_, ibase, _) when ibase < 2, do: {:error, "input base must be >= 2"}

  def convert(digits, input_base, output_base) do
    case do_convert_to_decimal(digits, input_base, 0) do
      {:error, _} = err -> err
      {:ok, decimal_value} -> do_convert_to_obase(decimal_value, output_base, [])
    end
  end

  @spec do_convert_to_decimal(
          digits :: [integer],
          input_base :: integer,
          decimal_value :: integer
        ) :: {:ok, integer} | {:error, String.t()}

  defp do_convert_to_decimal([], _, dec), do: {:ok, dec}

  defp do_convert_to_decimal([d | _], ibase, _) when d not in 0..(ibase - 1) do
    {:error, "all digits must be >= 0 and < input base"}
  end

  defp do_convert_to_decimal([d | digits], ibase, dec) do
    do_convert_to_decimal(digits, ibase, ibase * dec + d)
  end

  @spec do_convert_to_obase(
          decimal_value :: integer,
          output_base :: integer,
          digits :: [integer]
        ) :: {atom, [integer]}

  defp do_convert_to_obase(0, _, []), do: {:ok, [0]}
  defp do_convert_to_obase(0, _, digits), do: {:ok, digits}

  defp do_convert_to_obase(dec, obase, digits) do
    do_convert_to_obase(div(dec, obase), obase, [rem(dec, obase) | digits])
  end
end
